protocol-units/bridge/service/src/runtime.rs

```diff
@@ -0,0 +1,132 @@
+//use bridge_indexer_db::client::Client as IndexerClient;
+use bridge_util::{
+	actions::{ActionExecError, TransferAction, TransferActionType},
+	chains::bridge_contracts::BridgeContractEvent,
+	events::{InvalidEventError, TransferEvent},
+	states::TransferState,
+	types::BridgeTransferId,
+};
+use std::collections::HashMap;
+
+pub struct Runtime {
+	swap_state_map: HashMap<BridgeTransferId, TransferState>,
+}
+
+impl Runtime {
+	pub fn new() -> Self {
+		Runtime { swap_state_map: HashMap::new() } //indexer_db_client
+	}
+
+	pub fn iter_state(&self) -> impl Iterator<Item = &TransferState> {
+		self.swap_state_map.values()
+	}
+
+	pub fn remove_transfer(&mut self, transfer_id: BridgeTransferId) {
+		self.swap_state_map.remove(&transfer_id);
+	}
+
+	pub fn process_event<A>(
+		&mut self,
+		event: TransferEvent<A>,
+	) -> Result<TransferAction, InvalidEventError>
+	where
+		A: Into<Vec<u8>> + std::clone::Clone + std::fmt::Debug,
+	{
+		self.validate_state(&event)?;
+		let event_transfer_id = event.contract_event.bridge_transfer_id();
+		let state_opt = self.swap_state_map.remove(&event_transfer_id);
+		let (state, action) = match event.contract_event {
+			BridgeContractEvent::Initiated(detail) => {
+				TransferState::transition_from_initiated(event_transfer_id, detail)
+			}
+			BridgeContractEvent::Completed(_detail) => {
+				// Unwrap tested before in validate_state() state can be unwrap
+				let state = state_opt.unwrap();
```

Comment:
Why is unwrap appropriate here?

Reply:
All state are verified in the [validate_state()](https://github.com/movementlabsxyz/movement/blob/2c27679a4ae01a28a4ce38c71a76ba582d083602/protocol-units/bridge/service/src/runtime.rs#L35C8-L35C22) function before. You can't have a valid `BridgeContractEvent::Completed` event without an existing state so `self.swap_state_map.remove(&event_transfer_id)` will always return Some(state) for this type of event here.
It's tested [here](https://github.com/movementlabsxyz/movement/blob/2c27679a4ae01a28a4ce38c71a76ba582d083602/protocol-units/bridge/service/src/runtime.rs#L79).

Reply:
@musitdev @0xmovses I still don't think this is the right approach. It's quite easy to map and error here.